clippy_lints/src/module_style.rs

```diff
@@ -153,8 +153,11 @@ fn process_paths_for_mod_files<'a>(
 }
 
 /// Checks every path for the presence of `mod.rs` files and emits the lint if found.
+/// We should not emit a lint for test modules in the presence of `mod.rs`.
+/// Using `mod.rs` in integration tests is a [common pattern](https://doc.rust-lang.org/book/ch11-03-test-organization.html#submodules-in-integration-test)
+/// for code-sharing between tests.
 fn check_self_named_mod_exists(cx: &EarlyContext<'_>, path: &Path, file: &SourceFile) {
-    if path.ends_with("mod.rs") {
+    if path.ends_with("mod.rs") && !path.starts_with("tests") {
```

Comment:
There's a similar pattern on line 148, would it also need this check?

Reply:
I don't think it is needed. The pattern above collects the paths to the folders to be enforced when the old style is used.

The pattern here emits an error when the new style is used in the code but within the tests folder the old style with `mod.rs` is used.

I created a structure as in #11775 and the lint wasn't emitted.